internal/contour/route.go

```diff
@@ -234,12 +230,45 @@ func prefixmatch(prefix string) route.RouteMatch {
 
 // action computes the cluster route action, a *route.Route_route for the
 // supplied ingress and backend.
-func actionroute(namespace, name string, port int, ws bool, timeout time.Duration) *route.Route_Route {
-	cluster := hashname(60, namespace, name, strconv.Itoa(port))
+func actionroute(be []*dag.Service, ws bool, timeout time.Duration) *route.Route_Route {
+	totalWeight := 0
+	upstreams := []*route.WeightedCluster_ClusterWeight{}
+
+	// Loop over all the upstreams and add to slice
+	for _, i := range be {
+
+		// Create the empty upstream
+		upstream := route.WeightedCluster_ClusterWeight{
+			Name:   hashname(60, i.Namespace(), i.Name(), strconv.Itoa(int(i.Port))),
+			Weight: &types.UInt32Value{Value: uint32(0)},
+		}
+
+		// If weight is passed, then use otherwise calculate
+		if i.Weight != nil {
```

Comment:
We could modify dag.recompute so that weight was always non nil (see my comment below) and the weight value was always valid, ie more than zero. 

Then totalweight would never be zero and we wouldn't have to do a fixup.